calicoctl/commands/migrate/import.go

```diff
@@ -0,0 +1,409 @@
+}
+
+func importCRDs(cfg *apiconfig.CalicoAPIConfig) error {
+	// TODO: Add the apiextensions clientset and handling logic to the libcalico-go kubeclient
+	// Start a kube client
+	// Use the kubernetes client code to load the kubeconfig file and combine it with the overrides.
+	configOverrides := &clientcmd.ConfigOverrides{}
+	var overridesMap = []struct {
+		variable *string
+		value    string
+	}{
+		{&configOverrides.ClusterInfo.Server, cfg.Spec.K8sAPIEndpoint},
+		{&configOverrides.AuthInfo.ClientCertificate, cfg.Spec.K8sCertFile},
+		{&configOverrides.AuthInfo.ClientKey, cfg.Spec.K8sKeyFile},
+		{&configOverrides.ClusterInfo.CertificateAuthority, cfg.Spec.K8sCAFile},
+		{&configOverrides.AuthInfo.Token, cfg.Spec.K8sAPIToken},
+	}
+
+	// Set an explicit path to the kubeconfig if one
+	// was provided.
+	loadingRules := clientcmd.ClientConfigLoadingRules{}
+	if cfg.Spec.Kubeconfig != "" {
+		loadingRules.ExplicitPath = cfg.Spec.Kubeconfig
+	}
+
+	// Using the override map above, populate any non-empty values.
+	for _, override := range overridesMap {
+		if override.value != "" {
+			*override.variable = override.value
+		}
+	}
+	if cfg.Spec.K8sInsecureSkipTLSVerify {
+		configOverrides.ClusterInfo.InsecureSkipTLSVerify = true
+	}
+
+	// A kubeconfig file was provided.  Use it to load a config, passing through
+	// any overrides.
+	config, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+		&loadingRules, configOverrides).ClientConfig()
+	if err != nil {
+		return err
+	}
+
+	cs, err := clientset.NewForConfig(config)
+	if err != nil {
+		return err
+	}
+	log.Debugf("Created k8s CRD ClientSet: %+v", cs)
+
+	// Apply the CRDs
+	for _, crd := range crds.CalicoCRDs {
+		_, err := cs.ApiextensionsV1beta1().CustomResourceDefinitions().Create(crd)
+		if err != nil {
+			if kerrors.IsAlreadyExists(err) {
+				// If the CRD already exists attempt to update it.
+				// Need to retrieve the current CRD first.
+				currentCRD, err := cs.ApiextensionsV1beta1().CustomResourceDefinitions().Get(crd.GetObjectMeta().GetName(), v1.GetOptions{})
+				if err != nil {
+					return fmt.Errorf("Error retrieving existing CRD to update: %s: %s", crd.GetObjectMeta().GetName(), err)
+				}
+
+				// Use the resource version so that the current CRD can be overwritten.
+				crd.GetObjectMeta().SetResourceVersion(currentCRD.GetObjectMeta().GetResourceVersion())
+
+				// Update the CRD.
+				_, err = cs.ApiextensionsV1beta1().CustomResourceDefinitions().Update(crd)
+				if err != nil {
+					return fmt.Errorf("Error updating CRD %s: %s", crd.GetObjectMeta().GetName(), err)
+				}
+			} else {
+				return fmt.Errorf("Error creating CRD %s: %s", crd.GetObjectMeta().GetName(), err)
+			}
+		}
+		log.Debugf("Applied %s CRD", crd.GetObjectMeta().GetName())
+	}
+
+	return nil
+}
+
+func applyV3(args map[string]interface{}) error {
+	results := common.ExecuteConfigCommand(args, common.ActionApply)
+	log.Infof("results: %+v", results)
+
+	if results.FileInvalid {
```

Comment:
This feels like a lot of duplication in code. 

Could we instead keep these fields private on the results structure, and just commonize it via a method?

e.g., 

```
if err := results.Error(); err != nil {
  return err
}
```

Reply:
To be honest, I don't think it really matters much to the user if some resources applied and some didn't for this use-case anyway. I think it's a boolean - did it work? Or did it fail?

So, we can reduce some of complexity of tracking that and just return an error or not.

Becasue, if they hit an error, they need to clear and start over again anyway.

Reply:
A lot of the separate handling code differs slightly depending on the command it is being used from. The only way to keep the fields private would be to add separate `Error` functions i.e. `ErrorGetOrList`, `ErrorApply`, `ErrorCreate`, etc. which I opted against since the only truly duplicate piece of code is this invocation here and the one used by Apply.

Reply:
I shortened a lot of the error handling since, as you pointed out, it's not really necessary for this use case.